add option to limit domains on AddRedirectToWww

delebru opened the pull request:
@jkotalik: This is the idea I had in mind, let me know if you have any thoughts.

Addresses #9274

Commit:
add option to limit domains on AddRedirectToWww

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/aspnet/AspNetCore?pullRequest=9601) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/aspnet/AspNetCore?pullRequest=9601) before we can accept your contribution.<br/><br/>:x: delebru [sign now](https://cla.dotnetfoundation.org/aspnet/AspNetCore?pullRequest=9601)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/aspnet/AspNetCore?pullRequest=9601) it.</sub>

jkotalik commented:
This should target the master branch. I don't think we would consider adding a feature to a patch like this.

jkotalik commented on src/Middleware/Rewrite/src/Internal/RedirectToWwwRule.cs:
```diff
@@ -32,6 +44,12 @@ public virtual void ApplyRule(RewriteContext context)
                 return;
             }
+            
+            if (_domains.Length > 0 && !_domains.Any(d => req.Host.Host.Equals(d, StringComparison.OrdinalIgnoreCase)))
```
Few performance ideas:
- I'd throw in the constructor if domains.Length == 0. That way, you will always have at least one domain.
- Using LINQ here seems inefficient, it's probably more performant to just have a foreach loop through each domain.

delebru replied:
Thanks for the review. I just pushed a commit adding the throw and replacing linq with a foreach. For further improvements we could replace:
```foreach (var domain in _domains)
{
    if (domain.Equals(req.Host.Host, StringComparison.OrdinalIgnoreCase))
    {
        domainRequiresRedirection = true;
        break;
    }
}
```
with:
```for (var i = 0; i < _domains.Length; i++)
{
    if (_domains[i].ReferenceEquals(req.Host.Host) || _domains[i].Equals(req.Host.Host, StringComparison.OrdinalIgnoreCase))
    {
        domainRequiresRedirection = true;
        break;
    }
}
```
Let me know what you'd like to see that change commited.

jkotalik replied:
I don't think _domains[i].ReferenceEquals(req.Host.Host) would ever be true, as the domain you are passing into the rule wouldn't be the same string passed into the WwwRule.

delebru replied:
Oh you're right! The strings will on most cases come from different instances.

jkotalik commented on src/Middleware/Rewrite/src/Internal/RedirectToWwwRule.cs:
```diff
@@ -12,11 +13,22 @@ namespace Microsoft.AspNetCore.Rewrite.Internal
     public class RedirectToWwwRule : IRule
     {
         public readonly int _statusCode;
+        public string[] _domains = new string[0];
```
You don't need to set this to a default value if you take my suggestion on throwing if the domains is an empty array.

delebru replied:
I initialized the domains array with a default value in case the original constructor is called where we can't throw.

If that's a "deal breaker" we could try a different approach by adding another boolean that defines if the rule should apply to all domains. We can set that boolean to true on the original constructor and use it to skip the check on the domains array when applying the rule: 

```
public class RedirectToWwwRule : IRule
{
    public readonly int _statusCode;
    public readonly string[] _domains;
    public readonly bool _applyToAllDomains;

    public RedirectToWwwRule(int statusCode)
    {
        _statusCode = statusCode;
        _applyToAllDomains = true;
    }
        
    public RedirectToWwwRule(int statusCode, params string[] domains)
    {
        if (domains == null || domains.Length == 0)
        {
            throw new ArgumentNullException(nameof(domains));
        }
        if (domains.Length < 1)
        {
            throw new ArgumentException(nameof(domains));
        }
        _domains = domains;
        _applyToAllDomains = false;
        _statusCode = statusCode;
    }

    public virtual void ApplyRule(RewriteContext context)
    {
        var req = context.HttpContext.Request;
        if (req.Host.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase))
        {
            context.Result = RuleResult.ContinueRules;
            return;
        }

        if (req.Host.Value.StartsWith("www.", StringComparison.OrdinalIgnoreCase))
        {
            context.Result = RuleResult.ContinueRules;
            return;
        }

        if (!_applyToAllDomains)
        {
            var domainRequiresRedirection = false;
            foreach (var domain in _domains)
            {
                if (domain.Equals(req.Host.Host, StringComparison.OrdinalIgnoreCase))
                {
                    domainRequiresRedirection = true;
                    break;
                }
            }
            if (!domainRequiresRedirection)
            {
                context.Result = RuleResult.ContinueRules;
                return;
            }
        }

        var wwwHost = new HostString($"www.{req.Host.Value}");
        var newUrl = UriHelper.BuildAbsolute(req.Scheme, wwwHost, req.PathBase, req.Path, req.QueryString);
        var response = context.HttpContext.Response;
        response.StatusCode = _statusCode;
        response.Headers[HeaderNames.Location] = newUrl;
        context.Result = RuleResult.EndResponse;
        context.Logger?.RedirectedToWww();
    }
}
```

jkotalik replied:
Do that except with a null check instead:

```c#
public class RedirectToWwwRule : IRule
{
    public readonly int _statusCode;
    public readonly string[] _domains;

    public RedirectToWwwRule(int statusCode)
    {
        _statusCode = statusCode;
    }

    public RedirectToWwwRule(int statusCode, params string[] domains)
    {
        if (domains == null)
        {
            throw new ArgumentNullException(nameof(domains));
        }
        _domains = domains;
        _statusCode = statusCode;
    }

    public virtual void ApplyRule(RewriteContext context)
    {
        var req = context.HttpContext.Request;
        if (req.Host.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase))
        {
            context.Result = RuleResult.ContinueRules;
            return;
        }

        if (req.Host.Value.StartsWith("www.", StringComparison.OrdinalIgnoreCase))
        {
            context.Result = RuleResult.ContinueRules;
            return;
        }

//      if (_domains != null)
//      {
            var requestRequiresRedirection = false;
            foreach (var domain in _domains)
            {
                if (domain.Equals(req.Host.Host, StringComparison.OrdinalIgnoreCase))
                {
                    requestRequiresRedirection = true;
                    break;
                }
            }
            if (!requestRequiresRedirection)
            {
                context.Result = RuleResult.ContinueRules;
                return;
            }
//      }

        var wwwHost = new HostString($"www.{req.Host.Value}");
        var newUrl = UriHelper.BuildAbsolute(req.Scheme, wwwHost, req.PathBase, req.Path, req.QueryString);
        var response = context.HttpContext.Response;
        response.StatusCode = _statusCode;
        response.Headers[HeaderNames.Location] = newUrl;
        context.Result = RuleResult.EndResponse;
        context.Logger?.RedirectedToWww();
    }
}
```

delebru replied:
Yes! How didn't I see that option... Thanks, I love this solution :)

Commit:
Improvements for review

Added parameter length check and replaced linq with foreach to loop the domains string array.

delebru commented:
> This should target the master branch. I don't think we would consider adding a feature to a patch like this.

That's my bad, didn't intend for this to be used as a patch, just a mistake on the PR target branch. I'll open a new PR to master once we have a preferred solution from the previous comments.

Commit:
variable rename

jkotalik commented on src/Middleware/Rewrite/src/Internal/RedirectToWwwRule.cs:
```diff
@@ -32,6 +47,21 @@ public virtual void ApplyRule(RewriteContext context)
+                    break;
+                }
+            }
+            if (!requestRequiresRedirection)
```
nit: newline.

jkotalik commented on src/Middleware/Rewrite/src/Internal/RedirectToWwwRule.cs:
```diff
@@ -12,11 +12,26 @@ namespace Microsoft.AspNetCore.Rewrite.Internal
+            if (domains.Length < 1)
+            {
+                throw new ArgumentException(nameof(domains));
+            }
```
nit: newline

jkotalik commented on src/Middleware/Rewrite/src/RewriteOptionsExtensions.cs:
```diff
@@ -147,5 +168,17 @@ public static RewriteOptions AddRedirectToWww(this RewriteOptions options, int s
+        public static RewriteOptions AddRedirectToWww(this RewriteOptions options, int statusCode, params string[] domains)
+        {
+            options.Rules.Add(new RedirectToWwwRule(statusCode, domains));
+            return options;
```
Make sure to add some tests!

Bartmax commented:
> I don't think _domains[i].ReferenceEquals(req.Host.Host) would ever be true, as the domain you are passing into the rule wouldn't be the same string passed into the WwwRule.

Not sure I follow,
`ReferenceEquals("string", "string")  // returns true`
I see huge improvement (10x) using ReferenceEquals

delebru commented:
I'm closing this PR now. I'll add the tests and all the suggestions from the latest reviews and open a new PR to master tonight. Thanks!
